File: 0004-valid-parentheses/solution.swift
class Solution {
    func isValid(_ s: String) -> Bool {
        var stack: [Character] = []
        
        let pairs: [Character: Character] = [
            ")": "(",
            "]": "[",
            "}": "{"
        ]
        for char in s {
            switch char {
            case "(", "[", "{":
                stack.append(char)
            case ")", "]", "}":
                if let expected = pairs[char] {
                    if stack.isEmpty {
                        return false
                    }
                    let last = stack.removeLast()
                    if last != expected {
                        return false
                    }
                }
            default:
                return false
            }
        }
        return stack.isEmpty
    }
}

File: 0004-valid-parentheses/solution.go
package main

func isValid(s string) bool {
	stack := []rune{}
	mapping := map[rune]rune{
		')': '(',
		'}': '{',
		']': '[',
	}

	for _, char := range s {
		switch char {
		case '(', '{', '[':
			stack = append(stack, char)
		case ')', ']', '}':
			if len(stack) == 0 {
				return false
			}

			last := stack[len(stack)-1]
			stack = stack[:len(stack)-1]

			if last != mapping[char] {
				return false
			}
		default:
			return false
		}
	}
	return len(stack) == 0
}
